schema/field.go

```diff
@@ -83,18 +93,22 @@ type FieldSerializer interface {
 	Serialize(value interface{}) (interface{}, error)
 }
 
-// Compile implements Compiler interface and recursively compile sub schemas and
-// validators when they implement Compiler interface.
-func (f Field) Compile() error {
+// Compile implements the ReferenceCompiler interface and recursively compile sub schemas
+// and validators when they implement Compiler interface.
+func (f Field) Compile(rc ReferenceChecker) error {
 	// TODO check field name format (alpha num + _ and -).
 	if f.Schema != nil {
 		// Recursively compile sub schema if any.
-		if err := f.Schema.Compile(); err != nil {
+		if err := f.Schema.Compile(rc); err != nil {
 			return fmt.Errorf(".%v", err)
 		}
 	} else if f.Validator != nil {
-		// Compile validator if it implements Compiler interface.
-		if c, ok := f.Validator.(Compiler); ok {
+		// Compile validator if it implements the ReferenceCompiler or Compiler interface.
+		if c, ok := f.Validator.(ReferenceCompiler); ok {
+			if err := c.Compile(rc); err != nil {
+				return fmt.Errorf(": %v", err)
+			}
+		} else if c, ok := f.Validator.(Compiler); ok {
```

Comment:
The `Compiler` interface is also used within the resource package, e.g. for the index.

I can check if it's feasible to move the `Compiler` interface to the `resource` package, and only use the new `ReferenceCompiler` interface within the `schema` package. It's an even bigger change, especially when it commes to updating tests, but should be mostly search/replace.